### PR TITLE
Caldav server

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Add a CalDAV server endpoint in GLPI.

### CODE REVIEW

**Comment by cedric-anne** [comment]
There are some complicated parts and there is still a lot to do.

### ACL 
ACL are hardcoded in the used library and is is not easy to change as they are instanciated in a lot of places, sometimes far deep in the execution chain.
All objects served in a user calendar are considered as owned by this user, so user have all rights on it (from CalDAV point of view). So basically, a reminder from someone else served in user calendar may be updated or deleted by user itself.

Solutions:
 a. Override `CalendarObject` class (needs to override `CalendarRoot`, `CalendarHome` and `Calendar` to be able to control instanciation chain);
 b. Check rights when client tries to alter objects, and return an error if user has not enough rights.

As I do not even know if this ACL information is sent to the client, and if clients use them, I cannot tell if solution "a" is better than solution "b". Solution seems easier to implement, as it can be fully done onalready implemented backend class.

### Code factorization 
Factorization with planning view seems finally a bad idea, as we may need to server differents items and different properties.
For example, during my tests I saw that a client like Thunderbird handles tasks that are not planned, but they are not retrieve by the `populatePlanning` method.
Also, some properties are not used in the planning view but may be interesting to have in objetcs served by the CalDAV server (i.e. VALARM properties).

### Groups
I did not yet implement groups methods on principals backend, so it is not yet possible to retrieve calendars from groups members.

**Comment by cedric-anne** [comment]
Integrates #6146 for now to be able to work on external events creation from CalDAV client.

**Comment by cedric-anne** [comment]
ACL problems are now fixed and groups are impemented.

Following features may be added, but requires a lot of work and are not, IMHO, mandatory for this first version of CalDAV server:
 - usage of VALARM properties to sync "recalls",
 - Sync/Sharing/Subscription plugins support.

So I consider this PR ready for review/test.

File: inc/caldav/traits/caldavuriutiltrait.class.php [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+               'itemtype'
+            ],
+            'DISTINCT' => true,
+            'FROM'     => $union,
+         ]
+      );
+
+      if ($items_iterator->count() !== 1) {
+         // Ambiguous response, unable to return matching element.
+         // Should never happens as UID has very very low probability to not be unique.
```
Comment (trasher) [inline inc/caldav/traits/caldavuriutiltrait.class.php]: Maybe add a log entry

**Comment by cedric-anne** [comment]
I added lots of tests on more common cases, and fixes lots of small bugs.
PR can be reviewed and tested now.

File: inc/commonitiltask.class.php [hunk 2]
```diff
@@ -1970,4 +1983,126 @@ static function showVeryShort($ID, $itemtype) {
...
+      $item = new static();
+      $parent_item = getItemForItemtype($item->getItilObjectItemType());
+      if (!$parent_item) {
+         return;
+      }
+
+      $query = [
+         'SELECT'     => [$item->getTableField('*')],
+         'FROM'       => $item->getTable(),
+         'INNER JOIN' => [],
```
Comment (trasher) [inline inc/commonitiltask.class.php]: Does not an empty inner join cause issues?

Comment (cedric-anne) [reply inc/commonitiltask.class.php]: Nope, it executes a foreach on an empty array, resulting in appending an empty string to the query.

**Comment by orthagh** [comment]
Functionnal review

OK :
- autodiscover (good job)
- setup of available objects (VEVENT, VTODO, etc)

to do:
- add a icon discoverable in the interface -> add a breadcrumb icon (`fas fa-sync`)

**Comment by orthagh** [comment]
found error when moving a task in VEVENT mode!
```
Base component of VCALENDAR object must be a VTODO or a VJOURNAL[sabredav-version][exception][message]
```

**Pull request merged**